### ISSUE TITLE
Mapping of a Woo Order and a CiviCRM Contribution/Participant/Membership status

### ISSUE THREAD

**Issue by jbonlinea** [WPCV/wpcv-woo-civi-integration#11]
spliting #7 for better communication

> > Nevertheless, to my understanding if someone pass and pay an order for membership or participation, hes membership or participation should be "valid" immediately despite the contribution statut is "pending".
> 
> CiviCRM handles this via the Order API and Payment API. In my testing, CiviCRM does exactly what it should when the various parameters are set - i.e. Memberships and Participants are created and have their respective statuses set correctly. FYI, the requirement for a Price Field Value is to make CiviCRM behave itself.

ok I've dugg a bit deeper than previously and you'll find some more details below, but in short, when an order is paid by credit card (stripe test), the contribution is created, but not the payment record. This is thus normal that the contribution is marked as pending, ans that the membership/participation are not yet active.

However, this is the behaviour I expect for check or transfert paiement method, as indeed, the payment isn't done yet. It is not the behaviour I expect when paying online.

I hope you'll find this usefull :)

Cheers

---

### case 1 contribution
- product : civi contribution
- payment : immediate by stripe (test mode)
- woo-order statute : processing
- Civi contribution : pending (transaction incomplete) 
  - no payment recorded (see capture)

![Capture d’écran du 2021-11-23 13-56-14](https://user-images.githubusercontent.com/23659411/143027925-5f4189c7-37a9-4531-ac2f-f6b25626409d.png)

Expected behaviour :
- in civi contribution, a payment should be there

Ok then I switch the woo order to completed
- the contribution is marked as finished
  - the payment appears

![Capture d’écran du 2021-11-23 13-59-27](https://user-images.githubusercontent.com/23659411/143028401-7a5a29a2-dd79-4e52-80e5-7ed490f107fe.png)

Comments : 
- I don't mind if the payment.s are only synced once the woo order statut is completed/finished, rather than as soon as the paiement is done and the order maked as processing

---

### case 2 participation
- product : civi participant
- payment : immediate by stripe (test mode)
- woo-order statute : processing
- Civi contribution : pending (transaction incomplete) 
  - no payment recorded (see capture)
- civi participation : waiting (transaction incomplete)
![Capture d’écran du 2021-11-23 13-36-18](https://user-images.githubusercontent.com/23659411/143025159-c37d19db-3f55-444d-9edf-fe4df468d4c7.png)
![Capture d’écran du 2021-11-23 13-36-58](https://user-images.githubusercontent.com/23659411/143025184-43806b5b-d040-4fdd-9959-95c094981d2b.png)

expected behaviour (for me)
- in civi contribution, a payment should be there
- the participation statut should be registered/subscribed (in frecnch it is "inscrit" I turned Civi back to english but this remains in french)

Then in woo I change the statute to completed
- the contribution is marked as finished
  - a payment appears
- the participation is marked as "inscrit" in french (registered/subscribed in english)

Comments : 
- I don't mind if the payment.s are only synced once the woo order statut is completed/finished, rather than as soon as the paiement is done and the order maked as processing
-  I really expect the participation to be "completed" as soon as the payment is done, not after I change the order statut in woo.

---

#### case 3 membership
- product : civi membership
- payment : immediate by stripe (test mode)
- woo-order statute : processing
- Civi contribution : pending (transaction incomplete) 
  - no payment recorded (see capture)
- civi membership : pending and inactive -> pending

![Capture d’écran du 2021-11-23 14-03-40](https://user-images.githubusercontent.com/23659411/143029146-640d9402-391a-41ea-810b-529bb75746ad.png)

![Capture d’écran du 2021-11-23 14-03-55](https://user-images.githubusercontent.com/23659411/143029172-e1a5ed75-53b5-406f-ba6e-95b5036a63da.png)

expected behaviour (for me)
- in civi contribution, a payment should be there
- the membership statut should be curent (or new) 

Then in woo I change the statute to completed
- the contribution is marked as finished
  - a payment appears
- the memberhsip is curent (new)

Comments : 
- I don't mind if the payment.s are only synced once the woo order statut is completed/finished, rather than as soon as the paiement is done and the order maked as processing
-  I really expect the membership to be "curent" as soon as the payment is done, not after I change the order statut in woo.

---

#### case 4 : simple product with participation/membership
as above

---

#### more comments

In all above case, 

we observe : woo-processing <-> civi contribution transaction incomplete

while I would expect : woo on hold <-> civi contribution transaction incomplete

Actually I did a test : pass an order with a participation as above with stripe, and add the paiement in Civi (without changing anything in woo), the contribution was then market as finished and the participation completed.

I understand the need to differentiate processing (paid) and finished (paid and processed) in Woo, and why not in Civi, but if Civi do not has contribution statute to mimic this, then I prefer that the contribution to be completed (and the memberhsip/participation) rather than not, and track the processing of the order solely from woo

**Comment by christianwach** [comment]
> when an order is paid by credit card (stripe test), the contribution is created, but not the payment record

@jbonlinea My shared host was triggering `Out of memory` errors when the Stripe WebHook ran, leaving the Orders in the "Processing" state. Could this be your issue as well? #17 suggests that Stripe *does* work.

**Comment by jbonlinea** [comment]
Hi @christianwach 

Hum, I've only tried in stripe test mode, which is suppose to be good enough the say weather stipe do work of not.

When I make an order in any of the above cases and select stripe, the payment appears as successful in stripe. To my understanding I would say that stripe does work.

re-testing with wp-debug, I have nothing in the log but something that certainly do not affect woo and woo-civi

`[02-Dec-2021 15:46:58 UTC] PHP Notice:  ob_end_flush(): failed to send buffer of zlib output compression (1) in /home/www/wp-includes/functions.php on l$`

the server error.log have this but it do not seems to appears when I pass an order

`test.mysite.net [Thu Dec 02 16:39:36.015180 2021] [authz_core:error] [pid 28415:tid 140543174555392] [client MY.SERVER.IP.ADDRESS:0] AH01630: client denied by server configuration: /home/www/wp-content/uploads/civicrm/ConfigAndLog/CiviCRM.4e7d205f6cf8cab343405922.log
`

Let me know if I can dig anywhere else

regards

**Comment by christianwach** [comment]
> When I make an order in any of the above cases and select stripe, the payment appears as successful in stripe. To my understanding I would say that stripe does work.

@jbonlinea Yes, I get the same - but until the Stripe WebHook successfully calls WooCommerce, the Woo Order remains in the "Processing" state. Is that what you're finding - or are your Woo Orders "Completed"?

FYI the code in this plugin that calls the CiviCRM Payment API won't run until the Woo Order has the "Completed" status - so the CiviCRM Contribution  can't be completed (and therefore Membership and Participant statuses won't get updated) until the Stripe WebHook has run successfully and the Woo Order has the "Completed" status.

If your Woo Orders don't have the "Completed" status, I'd be looking at why that is first.

**Comment by jbonlinea** [comment]
Hi @christianwach 

All happens as describe above BUT :

when passing an order : 
- containing only a simple product, set as virtual and downloadable, and having Civi sync ON (which ever type) 
- paid with stripe (in test mode)

--> the order is marked as completed in Woo, and both the contribution and payment is created in Civi

**Comment by christianwach** [comment]
@jbonlinea So everything's working then?

**Comment by christianwach** [comment]
@jbonlinea Digging into this a bit further and it seems that it is standard behaviour for WooCommerce to leave Virtual Products in the "Processing" state. This can be changed as per the instructions here:

https://woocommerce.com/document/automatically-complete-orders/

You may prefer to use the `woocommerce_payment_complete` hook instead though - not everybody uses a Thank You page.

I'm going to close this issue because I think this plugin isn't responsible for what you're seeing. Feel free to reopen if the solution in the link doesn't work for you.

**Comment by jbonlinea** [comment]
Hi @christianwach 

Sorry for my late reply, I couldn't make it earlier, since I had to do more testing of woosync to double check some stuff.

I did my best to document this issue but I'm not sure we understood each other at 100% :) 

This issue is about workflow, so this is debatable, but let let me advocate the idea I tried to rised . You may envision it diferently and that would be fine, I just want to make sure you  understood what I meant at first and now.

#### What "bothers me/not

What "bothers" me is not woo behaviour (which is debatable and editable as you mention), and it is not the "workingness/bugginess" of woosync, as I haven't identify a bug, sot it works.

What bothers me is the fact that an order that is paid in woo is not a sufficient condition to create the  contribution in civi, indeed 
- an order in woo, that is paid, and in statut is processing --> do not create the contribution in civi
- an order in woo, that is paid, and in statut is finished--> do create the contribution in civi

Actually what happens now is that if and order is marked as finished, a contribution is created ; this makes sens as finished in after paid within the path above.

Yet, to my understanding, if something is paid, it is paid, in both realm, woo and civi, and this is not the case.


#### Why do that bother me ? 

Order statut are within a path, as woo example here : 
![image](https://user-images.githubusercontent.com/23659411/146357277-dc13d1c4-8a48-46d0-a65a-631b825a8f9c.png)

conceptually : the curent behaviour bothers me because when an ordered is paid it shoiuld be marked so, no matter what else (wether the order is in processing or finished or whatever custom statut we may want to add that follows "paid")  ; this means that the contribution should be created too.

This is even more critical when as single order can :
- contain diferent product, including, single products, single virtual and downloadable products, CiviMemberhsip products, etc.
- contain a single or variable product or any other kind of product linked to a Civi Membership/Participation/Contribution

Why is it critical : if someone order and pay a single - not virtual and downloadable - product, linked to a Civi membership or participation. the woo order statut will be processing, but the contribution won't be created. Thus the perrson that passed the order won't be marked in Civi as a member, or as a participant... untill the woo order is processed.

now  If processing the woo order : 
- only means clicking on a button to change the statute, and you volontairly have not pined the porduct as virtual and downloadable to be able to do this (tiny/unique/useless/double check/whatever) processing 
- if the processing involve work, that may takes time, then you have two options : 
  - wait until the processing is done before changing the order statut :
     - you can use woo statut to tack wich order has been processed or not
     - your "client" Membership or Participation won't be recorded in civi, so you can't use civi for that matter
   - process the order immediatly : 
     - you can't use woo statut to track which order has been processed or not (it has to be marked finished, so you'll have to make a plan to track your order diferently)
     - you cant use civi to handle your membership or even as members and participant are up to date.


The idea here is that, for instance, if someone order and pays a mebership, his contribution is registered imediatly, his membership too, what ever else, for instance that the organisation has to do finish his order (notihng, send a member card, send him the paper version of its journal, etc.)


#### to conclude

okay, that was al ong one again :)

I hope it wans't to boring and unsetsandable.

Again, I this is how we envision workflow, my experience push me to advocate that membership/participation should be recorded as soon as paiement is done, despite some further processing mitght be needed (woo statut can be enough to track the task left, otherwisse civicase might be a good way to track complicated process)

Cheers

**Comment by christianwach** [comment]
@jbonlinea I think the current statuses are correct - CiviCRM Contributions *are* created when the Order is submitted, but they are in the "pending" state as they should be until the Order is completed. Participants and Memberships should not be flagged as "current" until the Order is completed.

As I said, the flow of statuses is configurable by you.

**Comment by jbonlinea** [comment]
> @jbonlinea I think the current statuses are correct - CiviCRM Contributions _are_ created when the Order is submitted, but they are in the "pending" state as they should be until the Order is completed. Participants and Memberships should not be flagged as "current" until the Order is completed.

hum ok, I got your point, and why you set it this way

in civi contiribution terms the path is : pending (awaiting paiement) - pending (partial paiement - completed

There is no "in between" statut fitting the concept of "processing" after paiment but before finished. 

This said, 
actually I miss-expressed myself above, because what happends if indeed that the contribution is created once the order is created, but the paiement of the contribution is not once the paiment is done in woo ; the paiement of the contribution if only created when the order it finished in woo.

The fact is in civi, once a contribution has is paid (not a partial paiement) it switch to completed, and indeed the contribution is completed. However, the contribution, may be one part of the "selling porcess", as the order tied to the contribution my require extra work. 

This is the whole diference between an order, which may involve crafting/shipping,  an a contribution witch is a "pure" paiement and nothing else, and it reflect the background idea of woo commerce, wich is online shop oriented, and civi, where it's non profi donation oriented.


Thank's for your replies

Cheers

**Comment by christianwach** [comment]
@jbonlinea I agree with what you're saying about conceptual mismatches - there are compromises that have to be made when linking two different models. You can find [the current status mapping here](https://github.com/WPCV/wpcv-woo-civi-integration/blob/14edd20d8e195ca7b1fde87875ba5af41280dc0a/includes/classes/class-woo-civi-contribution.php#L921). If you like, I can add a filter for you to modify the return?

**Comment by jbonlinea** [comment]
yes ! that is exactly what I tried to say at first 😄   and I lately mixed up the cotribution and paiement while I did well diferencite them in my first post 🙄 

and yes there is indeed comprimise to be made ! each use case may benefit from an approach or an other in term of statut maping, I have to say that I'm puzzled about the idea that paid on one side is not paid on the other, but this is within my vision of the path/workflow, but I have to recognise that I'm far from using all civi features, for instance case.

Having a filter seems like a great plus so anyone can map woo and civi stature at will, but I'm not 100% sure I'll deploy civi-woo yet so I can't let you built it "just for me", but I greatly appreciate the unvaluable (tremendous) and consistent effort you put into the whole civi-wp realm :)

**Comment by christianwach** [comment]
Filter available in e8b9b468a5307af0d60936e5eda4f4f1ea59f9e3
